services/facility_queue/facility_queue.py

```diff
@@ -0,0 +1,195 @@
+
+                    if json_response['finishtimestamp']:
+                        req.status = "Committing photometry to database"
+                        commit_photometry(
+                            json_response,
+                            altdata,
+                            followup_request.id,
+                            instrument.id,
+                            followup_request.requester.id,
+                        )
+                        req.status = 'complete'
+                        session.add(req)
+
+                    elif json_response['starttimestamp']:
+                        followup_request.status = f"Task is running (started at {json_response['starttimestamp']})"
+                        await self.put(req)
+                    else:
+                        followup_request.status = f"Waiting for job to start (queued at {json_response['timestamp']})"
+                        await self.put(req)
+                else:
+                    followup_request.status = f'error: {response.content}'
+
+                session.add(followup_request)
+                session.commit()
+
+            else:
+                raise ValueError(f'API for {instrument.name} unknown')
+
+            # Simulate time taken to service this request
+            await asyncio.sleep(3)
+
+            print(f"{req.method} request to [{req.endpoint}] completed")
+
+
+queue = FacilityQueue()
+
+
+class QueueHandler(tornado.web.RequestHandler):
+    def get(self):
+        self.set_header("Content-Type", "application/json")
+        self.write({"status": "success", "data": {"queue_length": queue.qsize()}})
+
+    async def post(self):
+
+        try:
+            data = tornado.escape.json_decode(self.request.body)
+        except json.JSONDecodeError:
+            self.set_status(400)
+            return self.write({"status": "error", "message": "Malformed JSON data"})
+
+        # validate data here and return 400 if invalid
+        try:
+            req = FacilityTransactionRequest(**data)
+        except TypeError:
+            self.set_status(400)
+            return self.write(
+                {
+                    "status": "error",
+                    "message": "Invalid arguments; cannot construct facility request",
+                }
+            )
+
+        for field in ('method', 'endpoint'):
+            if getattr(req, field) is None:
+                self.set_status(400)
+                return self.write(
+                    {
+                        "status": "error",
+                        "message": f"Missing request attribute `{field}`",
+                    }
+                )
+
+        session = Session()
+
+        current_req = session.execute(
+            sa.select(FacilityTransactionRequest).where(
+                FacilityTransactionRequest.followup_request_id
+                == req.followup_request_id
+            )
+        ).first()
+        if current_req is not None:
```

Comment:
Should we also add a check for duplicate requests with different IDs (i.e. poking at the same external URL)?

Reply:
Nice idea too.